url64d634b8f0_agdturner_generic_tgz-pJ8-uk_ac_leeds_ccg_andyt_generic_visualisation_charts_Generic_BarChartJ8/src/uk/ac/leeds/ccg/andyt/generic/io/Generic_StaticIO.java

```diff
@@ -0,0 +1,2156 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package uk.ac.leeds.ccg.andyt.generic.io;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.PrintWriter;
+import java.io.StreamTokenizer;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import uk.ac.leeds.ccg.andyt.generic.core.Generic_ErrorAndExceptionHandler;
+//import java.nio.file.S
+
+//TODO http://java.sun.com/docs/books/tutorial/essential/io/legacy.html#mapping
+// http://java.sun.com/docs/books/tutorial/essential/io/fileio.html
+/**
+ * Class of static methods for helping with reading and writing (primarily)
+ * to/from file.
+ */
+public class Generic_StaticIO {
+
+    public static final String CLASSNAME = "Generic_StaticIO";
+
+    /**
+     * Creates a new instance of ObjectIO
+     */
+    public Generic_StaticIO() {
+    }
+
+    public static TreeSet<String> recursiveFileList(File file) {
+        TreeSet<String> result = new TreeSet<String>();
+        if (file.isDirectory()) {
+            File[] files = file.listFiles();
+            for (File file1 : files) {
+                TreeSet<String> subresult = recursiveFileList(file1);
+                result.addAll(subresult);
+            }
+            return result;
+        } else {
+            result.add(file.toString());
+            return result;
+        }
+    }
+
+    public static TreeSet<String> recursiveFileList(File file, int depth) {
+        TreeSet<String> result = new TreeSet<String>();
+        if (depth != 0) {
+            if (file.isDirectory()) {
+                File[] files = file.listFiles();
+                for (File f : files) {
+                    TreeSet<String> subresult = recursiveFileList(f, depth - 1);
+                    result.addAll(subresult);
+                }
+                return result;
+            } else {
+                result.add(file.toString());
+                return result;
+            }
+        }
+        result.add(file.toString());
+        return result;
+    }
+
+    /**
+     * Write object to file
+     *
+     * @param object
+     * @param f
+     */
+    public static void writeObject(
+            Object object,
+            File f) {
+        try {
+            ObjectOutputStream oos;
+            oos = getObjectOutputStream(f);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *oos*. The resource is referenced at lines 105, 106. The resource is closed at line 107. There are other execution paths that do not contain closure statements, for example, when ObjectOutputStream.flush throws an exception. Ensure *oos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).